sync2/migration.go

```diff
@@ -0,0 +1,275 @@
+	}()
+	return sqlutil.WithTransaction(db, func(txn *sqlx.Tx) (err error) {
+		migrated, err := isMigrated(txn)
+		if err != nil {
+			return
+		}
+		if migrated {
+			logger.Debug().Msg("MigrateDeviceIDs: migration has already taken place")
+			return nil
+		}
+		logger.Info().Msgf("MigrateDeviceIDs: starting (commit=%t)", commit)
+
+		err = alterTables(txn)
+		if err != nil {
+			return
+		}
+
+		err = runMigration(txn, secret, whoamiClient)
+		if err != nil {
+			return
+		}
+		err = finish(txn)
+		if err != nil {
+			return
+		}
+
+		if !commit {
+			err = fmt.Errorf("MigrateDeviceIDs: migration succeeded without errors, but commit is false - rolling back anyway")
+		} else {
+			logger.Info().Msg("MigrateDeviceIDs: migration succeeded - committing")
+		}
+		return
+	})
+}
+
+func isMigrated(txn *sqlx.Tx) (migrated bool, err error) {
+	// Keep this dead simple for now. This is a one-off migration, before version 1.0.
+	// In the future we'll rip this out and tell people that it's the job to ensure this
+	// migration has ran before they upgrade beyond the rip-out point.
+	err = txn.QueryRow(`
+		SELECT EXISTS(
+		    SELECT 1 FROM information_schema.columns
+			WHERE table_name = 'syncv3_sync2_devices' AND column_name = 'user_id'
+		);
+	`).Scan(&migrated)
+
+	if err != nil {
+		err = fmt.Errorf("checkNotMigrated failed: %s", err)
+	}
+	return
+}
+
+func alterTables(txn *sqlx.Tx) (err error) {
+	_, err = txn.Exec(`
+		ALTER TABLE syncv3_sync2_devices
+		DROP CONSTRAINT syncv3_sync2_devices_pkey;
+	`)
+	if err != nil {
+		return
+	}
+
+	_, err = txn.Exec(`
+		ALTER TABLE syncv3_to_device_messages
+		DROP CONSTRAINT syncv3_to_device_messages_pkey,
+		ADD COLUMN user_id TEXT;
+	`)
+	if err != nil {
+		return
+	}
+
+	_, err = txn.Exec(`
+		ALTER TABLE syncv3_to_device_ack_pos
+		DROP CONSTRAINT syncv3_to_device_ack_pos_pkey,
+		ADD COLUMN user_id TEXT;
+	`)
+	if err != nil {
+		return
+	}
+
+	_, err = txn.Exec(`
+		ALTER TABLE syncv3_txns
+		DROP CONSTRAINT syncv3_txns_user_id_event_id_key,
+		ADD COLUMN user_id TEXT;
+	`)
+
+	return
+}
+
+type oldDevice struct {
+	AccessToken          string // not a DB row, but it's convenient to write to here
+	AccessTokenHash      string `db:"device_id"`
+	UserID               string `db:"user_id"`
+	AccessTokenEncrypted string `db:"v2_token_encrypted"`
+	Since                string `db:"since"`
+}
+
+func runMigration(txn *sqlx.Tx, secret string, whoamiClient Client) (err error) {
+	logger.Info().Msg("Loading old-style devices into memory")
+	var devices []oldDevice
+	err = txn.Select(
+		&devices,
+		`SELECT device_id, user_id, v2_token_encrypted, since FROM syncv3_sync2_devices;`,
+	)
+	if err != nil {
+		return
+	}
+
+	logger.Info().Msgf("Got %d devices to migrate", len(devices))
+
+	hasher := sha256.New()
+	hasher.Write([]byte(secret))
+	key := hasher.Sum(nil)
+
+	// TODO: can we get away with doing this sequentially, or should we parallelise
+	//       this like the poller startup routine does?
```

Comment:
This is slower but probably fine?

Reply:
Yeah. Just trying to keep things simple for now.

Reply:
On a test copy of the matrix.org deployment's database, this took <5min. I think that's more than fine.